### PR TITLE
add secp256k1 module

### PR DESCRIPTION
**Pull request by pkhry**
## Depends
 - [x] #249 
 
## Problem

We need to support tz2 inside interop, but we dont have secp256k1 module availible
## Solution

Add a secp256k1 module to crypto library
## Related
- #199

### CODE REVIEW

File: crypto/secp256k1.re [hunk 1]
```diff
@@ -0,0 +1,123 @@
+open Helpers;
+open Blake2b_helpers;
+
+module Libsecp256k1 = Libsecp256k1.External;
+open Libsecp256k1;
+let context = {
+  Mirage_crypto_rng_unix.initialize(); // TODO: remove when migrating to hacl-star
```
Comment (d4hines) [inline crypto/secp256k1.re:7]: This needs to be removed when #245 is merged

File: crypto/secp256k1.re [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  let of_raw = BLAKE2B_20.of_raw_string;
+  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+  let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+};
+module Secret = {
+  type t = Libsecp256k1.Key.t(Libsecp256k1.Key.secret);
+
+  let equal = (a, b) => {
+    Libsecp256k1.Key.equal(a, b);
+  };
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:66]:
```suggestion
  let equal = Libsecp256k1.Key.equal;
```

File: crypto/secp256k1.re [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    );
+  };
+  let _size = Libsecp256k1.Key.secret_bytes;
+  let prefix = Base58.Prefix.secp256k1_secret_key;
+  let to_raw = t =>
+    Libsecp256k1.Key.to_bytes(context, t) |> Bigstring.to_string;
+  let of_raw = string => {
+    Libsecp256k1.Key.read_sk(context, string |> Bigstring.of_string)
+    |> Result.to_option;
+  };
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:80]:
```suggestion
  let of_raw = string =>
    string
    |> Bigstring.of_string
    |> Libsecp256k1.Key.read_sk(context)
    |> Result.to_option;
```

File: crypto/secp256k1.re [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+  let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+};
+module Signature = {
+  type t = Sign.t(Sign.plain);
+
+  let equal = (a, b) => {
+    Sign.equal(a, b);
+  };
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:90]:
```suggestion
  let equal = Sign.equal;
```

File: crypto/secp256k1.re [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    Bigstring.compare(Sign.buffer(a), Sign.buffer(b));
+  };
+  let _size = Sign.plain_bytes;
+  let prefix = Base58.Prefix.secp256k1_signature;
+  let to_raw = t =>
+    Sign.to_bytes(~der=false, context, t) |> Bigstring.to_string;
+  let of_raw = string =>
+    Sign.read(context, string |> Bigstring.of_string) |> Result.to_option;
+  let to_string = (t: t): string =>
+    Base58.simple_encode(~prefix, ~to_raw, t);
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:101]:
```suggestion
  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
```

File: crypto/secp256k1.re [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  let equal = (a, b) => {
+    Sign.equal(a, b);
+  };
+  let compare = (a, b) => {
+    Bigstring.compare(Sign.buffer(a), Sign.buffer(b));
+  };
+  let _size = Sign.plain_bytes;
+  let prefix = Base58.Prefix.secp256k1_signature;
+  let to_raw = t =>
+    Sign.to_bytes(~der=false, context, t) |> Bigstring.to_string;
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:97]: what is `der`?

File: crypto/secp256k1.re [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  let of_raw = string =>
+    Sign.read(context, string |> Bigstring.of_string) |> Result.to_option;
```
Comment (EduardoRFS) [inline crypto/secp256k1.re:99]:
```suggestion
    string |> Bigstring.of_string |> Sign.read(context) |> Result.to_option;
```

**Comment by callistonianembrace** [comment]
Is this an improved attempt of #220 ? @EduardoRFS What about your comment here: https://github.com/marigold-dev/deku/pull/220#issuecomment-946111843 - can the reviewers ignore the old PR?

File: crypto/secp256k1.re (hunk 6 above)
Comment (callistonianembrace) [reply crypto/secp256k1.re:97]: Encoding rules IIUC
https://datatracker.ietf.org/doc/html/rfc5915#ref-X.690

**Review by callistonianembrace — APPROVED** [review]
LGTM. My questions/requests are addressed by existing review comments too

**Comment by EduardoRFS** [comment]
@callistonianembrace yeah the old PR should be ignored here.

**Pull request merged**